tokio/src/sync/mutex.rs

```diff
@@ -86,11 +6,88 @@ use std::error::Error;
 use std::fmt;
 use std::ops::{Deref, DerefMut};
 
-/// An asynchronous mutual exclusion primitive useful for protecting shared data
+/// An asynchronous `Mutex`-like type.
 ///
-/// Each mutex has a type parameter (`T`) which represents the data that it is protecting. The data
-/// can only be accessed through the RAII guards returned from `lock`, which
-/// guarantees that the data is only ever accessed when the mutex is locked.
+/// This type acts similarly to an asynchronous [`std::sync::Mutex`], with one
+/// major difference: [`lock`] does not block.  
+/// Another difference is that the lock guard can be held across await points.
+/// This allows you to do something along the lines of:
+///
+/// ```rust,no_run
+/// use tokio::sync::Mutex;
+/// use std::sync::Arc;
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let data1 = Arc::new(Mutex::new(0));
+///     let data2 = Arc::clone(&data1);
+///
+///     tokio::spawn(async move {
+///         let mut lock = data2.lock().await;
+///         *lock += 1;
+///     });
+///
+///     let mut lock = data1.lock().await;
+///     *lock += 1;
+/// }
+/// ```
+///
+/// Another example
+/// ```rust,no_run
+/// # #![warn(rust_2018_idioms)]
+/// #
```

Comment:
```suggestion
```

This lint is already used in crate root, so we can remove it.

https://github.com/tokio-rs/tokio/blob/282b00cbe888a96669877ce70662fba87e8c0e3c/tokio/src/lib.rs#L14-L17